packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts

```diff
@@ -0,0 +1,1096 @@
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: contract2, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: l2Contract1, delay: 0 },
+          { address: l2Proxy, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: l2Contract2, delay: 0 },
+          { address: l2Proxy, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+    ])
+  })
+
+  it('zksync lite', () => {
+    const verifier = EthereumAddress.random()
+    const governance = EthereumAddress.random()
+    const zkSync = EthereumAddress.random()
+    const upgradeGatekeeper = EthereumAddress.random()
+    const zkSyncMsig = EthereumAddress.random()
+    const actorA = EthereumAddress.random()
+    const actorB = EthereumAddress.random()
+
+    const graph: Graph = {
+      nodes: [
+        { address: verifier, threshold: 1 },
+        { address: governance, threshold: 1 },
+        { address: zkSync, threshold: 2 },
+        { address: upgradeGatekeeper, threshold: 1 },
+        { address: zkSyncMsig, threshold: 1 },
+        { address: actorA, threshold: 1 },
+        { address: actorB, threshold: 1 },
+      ],
+      edges: [
+        { type: 'upgrade', fromNode: 0, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 1, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 2, toNode: 3, delay: 0 },
+        { type: 'act', fromNode: 3, toNode: 4, delay: 21 },
+        { type: 'configure', fromNode: 2, toNode: 2, delay: 0 },
+        { type: 'member', fromNode: 2, toNode: 5, delay: 0 },
+        { type: 'member', fromNode: 2, toNode: 6, delay: 0 },
+      ],
+    }
+
+    expect(resolvePermissions(graph)).toEqualUnsorted([
+      {
+        type: 'upgrade',
+        path: [
+          { address: verifier, delay: 0 },
+          { address: upgradeGatekeeper, delay: 21 },
+          { address: zkSyncMsig, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: governance, delay: 0 },
+          { address: upgradeGatekeeper, delay: 21 },
+          { address: zkSyncMsig, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: zkSync, delay: 0 },
+          { address: upgradeGatekeeper, delay: 21 },
+          { address: zkSyncMsig, delay: 0 },
+        ],
+      },
+      {
+        type: 'configure',
+        path: [
+          { address: zkSync, delay: 0 }, // contract
+          { address: zkSync, delay: 0 }, // embedded security council
+        ],
+      },
+    ])
+  })
+
```

Comment:
Tests to add:

A->A (zksync lite simplified)
A->A->B
A->B->A
A->B->A->C->A (ignore loop - a upgrade b, b act a, a act c, c act a)